### PR TITLE
Add ical subscription

### PR DESCRIPTION
**Pull request by david-venhoff**
### Short description

This pr implements support for automatically creating events from an icalendar. This only implements step 1 of issue #100, as it is not currently possible to automatically use machine translation with this feature and we need another pr for that. However, it is possible to manually start machine translation for imported events.

For now, this feature is only visible to service team members and cms team members.

To test this pr, you could use this ical url, which includes events for july from Saarburg: https://www.saarburg.de/?post_type=tribe_events&tribe-bar-date=2024-07-08&ical=1&eventDisplay=list

### Proposed changes

- Add a new model for external calendars, which contain an url to the ical file
- Add fields to the event model to specify whether this event was automatically imported, and link it to the icalendar
- Implement functionality for converting icalendar events into cms events and event translations
- Add UI for managing external calendars

### Side effects

- We need a new cronjob that runs the import events command every once in a while

### Resolved issues

Fixes: #100 

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

**Comment by JoeyStk** [comment]
This PR is still blocked until #2850 is merged, so we can fix pylint. But feel free to review this PR already anyway as the required changes will be minimal :)

**Comment by JoeyStk** [comment]
https://github.com/digitalfabrik/integreat-cms/pull/2850 is merged now :)

**Comment by david-venhoff** [comment]
Both blockers are now resolved :)

**Review by MizukiTemma — CHANGES_REQUESTED** [review]
Thank you so much 💪 It looks a good start 😻 

I have some suggestions but feel free to discard (or postpone if it is subject to planned extensions of the next step(s)).

1. Fixing width of the columns in the external calendar list (see screen shots below)
We are getting this kind of fix requests time to time. Maybe better to implement it already now?

2. Checking the format of input in the field "URL"
We have format check for the URL field of the model Organization. It probably makes sense here too?

3. Warning when the URL or tag of a calendar is about to be changed
Events that are imported from the calendar and already saved will be deleted after changing those fields, if they do not match the URL and tag anymore. In my opinion warning before saving makes the system less confusing and safe.

External calendar 1
External calendar 2

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (deen13) [inline integreat_cms/cms/utils/external_calendar_utils.py:1]: If this utility is only used in one place (such as in the command), I would argue that it is more business-specific code than a utility and should be inlined.

**Review by deen13 — COMMENTED** [review]
First off, great job on this pull request! Your effort in tackling these issues is much appreciated. 🥳

#### General thoughts on the feature

- **Prefetch Existing Tags**: If possible, it might be useful to fetch all tags once the domain is entered. This way, tags could be presented as a combobox instead of a text field, which could help prevent errors.
- **Import Error**: There is an error occurring when importing events from Saarburg. The specific error is: `Aug 01 08:46:13 ERROR integreat_cms.core.storages - An error occurred while importing events from this external calendar and Could not import 'Ausstellung Günter Schuster: Material und Sprache': The maximum duration for events is 28 days. Consider using recurring events if the event is not continuous.`
- **Django Cron**: Pull Request #2920 introduces `django_cron`, which could be beneficial for our use case.
- **Event Review**: It would be helpful to have an overview of all imported and erroneous events after import and before storing them. This would facilitate a thorough review process.
- **Error Handling**: The status column provides a detailed explanation of errors, but it's unclear how to handle them. Consider adding a popup for errors to provide more space if there are multiple issues.
- **Error Management**: Adding a button to clear reviewed errors would be useful so that subsequent users do not have to deal with errors that have already been addressed. This would also aid in identifying new errors during cron job polling.
- **Security Considerations**: Should we implement some form of signing for this feature to prevent domain takeover and malicious event injection? [More on domain takeover attacks](https://claudijd.github.io/2017/02/03/deep-thoughts-on-subdomain-takeovers/)

File: integreat_cms/cms/migrations/0098_external_calendar.py
Comment (david-venhoff) [reply integreat_cms/cms/migrations/0098_external_calendar.py:34]: This code is copied from a few older migrations, which use this pattern to update the migrations. It might be worth to extract that into a common function though.

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (david-venhoff) [reply integreat_cms/cms/utils/external_calendar_utils.py:1]: It is not though, take a look at [here](https://github.com/digitalfabrik/integreat-cms/pull/2895/files#diff-062b555c7f2c89409dc7a6f9009039445dca3ba7e98725bad94147608acf4cb2R114) for example.

**Comment by david-venhoff** [comment]
Thanks for your review!

> Prefetch Existing Tags

This seems like a nice QoL improvement. However, that would not be a part of this pr, since we currently just intend to provide an mvp for the service team to test this feature. If everything works out, we will build on this mvp later.

> Import error

I don't think there is much we can do about that. It has been a design decision to not allow events longer than 28 days in our system and the external calendar contains an event that is longer than that.

> Django Cron

I didn't have a look at the pr yet, but that could be a nice alternative :)

> Event Review

This could be a separate issue that we implement later

> Error Handling

Same here

> Error Management

Same here

> Security Considerations

What is the threat you are thinking of? Are you worried that municipalities redirect to another provider for providing their icalendars? Because then I think it would be up to the municipalities to make sure that there are no security issues. 
Or that municipalities enter the url of a third-party website, which might be malicious? That does not seem like a domain takeover vulnerability to me and also seems like it would be the responsibility of the municipalities to me.

**Comment by JoeyStk** [comment]
> Warning when the URL or tag of a calendar is about to be changed
Events that are imported from the calendar and already saved will be deleted after changing those fields, if they do not match the URL and tag anymore. In my opinion warning before saving makes the system less confusing and safe.

I think this also would probably be something for a new issue :)

**Comment by MizukiTemma** [comment]
> > Warning when the URL or tag of a calendar is about to be changed
> > Events that are imported from the calendar and already saved will be deleted after changing those fields, if they do not match the URL and tag anymore. In my opinion warning before saving makes the system less confusing and safe.
> 
> I think this also would probably be something for a new issue :)

Ok 😉 Will you open an issue for that?

**Comment by JoeyStk** [comment]
> Ok 😉 Will you open an issue for that?

Sure! It's the first comment [in here](https://github.com/digitalfabrik/integreat-cms/issues/2967) :)

File: integreat_cms/cms/migrations/0098_external_calendar.py
Comment (david-venhoff) [reply integreat_cms/cms/migrations/0098_external_calendar.py:34]: I have added #2996 to address this now.

File: integreat_cms/cms/models/external_calendars/external_calendar.py
Comment (PeterNerlich) [inline integreat_cms/cms/models/external_calendars/external_calendar.py:28]: The iCalendar specification speaks of categories, not tags. I would like to at least let our internal terminology agree with the official one. The only exception I can think of would be if we already had event categories as a concept in integreat that couldn't overlap with this.
This is the only thing I would rather prefer doing in this PR, as we would need to create a data migration in the next PR otherwise.

```suggestion
    import_filter_category = models.CharField(
        max_length=255,
        blank=True,
        default=settings.EXTERNAL_CALENDAR_CATEGORY,
```

File: integreat_cms/cms/models/external_calendars/external_calendar.py
Comment (PeterNerlich) [inline integreat_cms/cms/models/external_calendars/external_calendar.py:31]: I wonder whether we should rename this also for the user or if *tags* is originating somewhere which would make this more familiar and intuitive to the user…?

```suggestion
        verbose_name=_(
            "The category that events need to have to get imported (Leave blank to import all events)"
        ),
    )
```

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (PeterNerlich) [inline integreat_cms/cms/utils/external_calendar_utils.py:24]: This might cause confusion over whether we long time *store* this. We don't at least not in this form, only when it was put into an Event.
```suggestion
    Holds data extracted from ical events
```

File: integreat_cms/cms/utils/external_calendar_utils.py
Comment (PeterNerlich) [inline integreat_cms/cms/utils/external_calendar_utils.py:151]: Seeing so many sands at the beach made me thirsty, I drank many water.
```suggestion
            _("The data provided by the url of this external calendar is invalid")
```

File: tests/cms/views/events/external_calendar.py
Comment (PeterNerlich) [inline tests/cms/views/events/external_calendar.py:45]: Shouldn't be this rather be the POST request submitting data? We don't really care about someone getting to an empty form rather than actually managing to submitting malicious data

**Review by PeterNerlich — APPROVED** [review]
This is mostly nitpicks and nothing serious blocking this PR, especially since this is more like just a prove of concept to see if we move forward with this idea.

Another such nitpick is that the tests cover the command itself and the permissions for the views, but not that the view actually triggers the same functionality.

File: integreat_cms/core/signals/feedback_signals.py
Comment (david-venhoff) [reply integreat_cms/core/signals/feedback_signals.py:74]: No, this code was originally part of #2781, but did not end up being included for merging, because it was not required anymore. This exception cannot occur anymore and the code should be removed from develop imo

Comment (JoeyStk) [reply integreat_cms/core/signals/feedback_signals.py:74]: Oh. Could you please open up a separate PR for that? I had assumed this was just a clean-up

Comment (david-venhoff) [reply integreat_cms/core/signals/feedback_signals.py:74]: #3038 :smile:

File: integreat_cms/static/src/css/style.scss
Comment (david-venhoff) [reply integreat_cms/static/src/css/style.scss:337]: Note that this css rule does not apply to all tables. Instead, it only applies to tables which are children of a node with a table-listing class (see here: https://github.com/digitalfabrik/integreat-cms/blob/88a67cdac9e5e7313ba2b175712ee601106b9c0d/integreat_cms/static/src/css/style.scss#L321)

Of those tables, all have the `w-full` class.

Unfortunately, the change still caused an error, because it accidentally fixed another bug, which caused the poi table rendering to go wrong :sob: 
I have opened a pr to fix that at #3044

Comment (PeterNerlich) [reply integreat_cms/static/src/css/style.scss:337]: oh, I forgot that this is ***S***CSS when I saw this :sweat_smile:

**Pull request merged**
